### ISSUE TITLE
ReportErrors is not thread safe but can be invoked concurrently with /parallel

### ISSUE THREAD

**Issue by nguerrera** [dotnet/roslyn#12550]
For the most part, calls to`CommonCompiler.ReportErrors` do not happen concurrently. However, there is one call in each of VB and C#'s `ParseFile` when there is an I/O error that can be invoked concurrently in the `/parallel` case. There are two unsafe operations with invoking `CommonCompiler.ReportErrors` concurrently:

1. Updating `HashSet<Diagnostic> _reportedDiagnostics`,
2. Calling `ErrorLogger.LogDiagnostic` for `/errorlog`

So far, I have only managed to construct a repro (detailed below) that illustrates (2).

**Version Used**: 1.3.1.60616 (still present in 2.0 by code inspection of master)

**Steps to Reproduce**:

1.  Create two empty files: a.cs and b.cs
2.  Change permissions of both files to deny all access
3.  `csc a.cs b.cs /parallel /errorlog:log.json`
4. Repeat 1-3 several times

**Expected Behavior**:
A valid error log is produced.

**Actual Behavior**:
It sometimes works (concurrency issue), and sometimes fails in at least two different ways: 

1. A malformed error log is silently produced.
2. An unhandled exception is raised:
```
Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.CodeAnalysis.ErrorLogger.DiagnosticDescriptorSet.Add(DiagnosticDescriptor descriptor)
   at Microsoft.CodeAnalysis.ErrorLogger.LogDiagnostic(Diagnostic diagnostic)
   at Microsoft.CodeAnalysis.CommonCompiler.ReportErrors(IEnumerable`1 diagnostics, TextWriter consoleOutput, ErrorLogger errorLoggerOpt)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompiler.ParseFile(TextWriter consoleOutput, CSharpParseOptions parseOptions, CSharpParseOptions scriptParseOptions, Boolean& hadErrors, CommandLineSourceFile file, ErrorLogger errorLogger, String& normalizedFilePath)
```
